fabtests/unit/ep_test.c

```diff
@@ -0,0 +1,444 @@
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <string.h>
+#include <time.h>
+
+#include <rdma/fi_errno.h>
+
+#include "unit_common.h"
+#include "shared.h"
+
+#define EP_TEST_SIZES_STRIDE 101
+
+static unsigned int num_eps, max_eps;
+struct fid_ep **ep_arr;
+struct fid_av **av_arr;
+struct fid_cq **txcq_arr, **rxcq_arr;
+static unsigned int seed = -1;
+
+static char err_buf[512];
+
+static int create_av(struct fid_av **av, enum fi_av_type type, int rx_ctx_bits,
+		     size_t count, size_t ep_per_node, const char *name,
+		     void *map_addr, uint64_t flags)
+{
+	struct fi_av_attr av_attr;
+
+	memset(&cq_attr, 0, sizeof(av_attr));
+	av_attr.type = type;
+	av_attr.rx_ctx_bits = rx_ctx_bits;
+	av_attr.count = count;
+	av_attr.ep_per_node = ep_per_node;
+	av_attr.name = name;
+	av_attr.map_addr = map_addr;
+	av_attr.flags = flags;
+
+	return fi_av_open(domain, &av_attr, av, NULL);
+}
+
+static int create_cq(struct fid_cq **cq, size_t size, uint64_t flags,
+		     enum fi_cq_format format, enum fi_wait_obj wait_obj)
+{
+	struct fi_cq_attr cq_attr;
+
+	memset(&cq_attr, 0, sizeof(cq_attr));
+	cq_attr.size = size;
+	cq_attr.flags = flags;
+	cq_attr.format = format;
+	cq_attr.wait_obj = wait_obj;
+
+	return fi_cq_open(domain, &cq_attr, cq, NULL);
+}
+
+static void close_ep(int idx)
+{
+	FT_CLOSE_FID(ep_arr[idx]);
+	FT_CLOSE_FID(av_arr[idx]);
+	FT_CLOSE_FID(txcq_arr[idx]);
+	FT_CLOSE_FID(rxcq_arr[idx]);
+}
+
+static int open_ep(int idx)
+{
+	int ret;
+
+	struct fid_ep **ep   = &ep_arr[idx];
+	struct fid_av **av   = &av_arr[idx];
+	struct fid_cq **txcq = &txcq_arr[idx];
+	struct fid_cq **rxcq = &rxcq_arr[idx];
+
+	fi->tx_attr->op_flags = 0;
+	ret = fi_endpoint(domain, fi, ep, NULL);
+	if (ret) {
+		FT_UNIT_STRERR(err_buf, "fi_endpoint failed when opening new EP", ret);
+		return ret;
+	}
+
+	ret = create_av(av, FI_AV_UNSPEC, 0, fi->domain_attr->ep_cnt,
+			fi->domain_attr->ep_cnt, NULL, NULL, 0);
+	if (ret)
+		goto out_close_ep;
+	ret = create_cq(txcq, fi->tx_attr->size, 0, FI_CQ_FORMAT_UNSPEC, FI_WAIT_NONE);
+	if (ret)
+		goto out_close_ep;
+	ret = create_cq(rxcq, fi->rx_attr->size, 0, FI_CQ_FORMAT_UNSPEC, FI_WAIT_NONE);
+	if (ret)
+		goto out_close_ep;
+
+	FT_EP_BIND(*ep, *av, 0);
+	FT_EP_BIND(*ep, *txcq, FI_TRANSMIT);
+	FT_EP_BIND(*ep, *rxcq, FI_RECV);
+	ret = fi_enable(*ep);
```

Comment:
The above code could probably call ft_enable_ep().  For MSG EP, we also need an EQ.